src/Namecheap/Provider.php

```diff
@@ -0,0 +1,502 @@
+        $tld = Utils::normalizeTld($params->tld);
+        $domainName = Utils::getDomain($sld, $tld);
+
+        $this->checkRegisterParams($params);
+
+        $checkResult = $this->api()->checkMultipleDomains($domainName);
+
+        if (count($checkResult) < 1) {
+            throw $this->errorResult('Empty domain availability check result');
+        }
+
+        if (!$checkResult[0]->can_register) {
+            throw $this->errorResult('This domain is not available to register');
+        }
+
+        $contacts = [
+            NamecheapApi::CONTACT_TYPE_REGISTRANT => $params->registrant->register,
+            NamecheapApi::CONTACT_TYPE_ADMIN      => $params->admin->register,
+            NamecheapApi::CONTACT_TYPE_TECH       => $params->tech->register,
+            NamecheapApi::CONTACT_TYPE_BILLING    => $params->billing->register,
+        ];
+
+        try {
+            $this->api()->register(
+                $domainName,
+                intval($params->renew_years),
+                $contacts,
+                $this->prepareNameservers($params, 'nameservers.ns'),
+            );
+
+            return $this->_getInfo($domainName, sprintf('Domain %s was registered successfully!', $domainName));
+        } catch (\Throwable $e) {
+            $this->handleException($e, $params);
+        }
+    }
+
+    /**
+     * @param  RegisterDomainParams  $params
+     *
+     * @return void
+     */
+    private function checkRegisterParams(RegisterDomainParams $params): void
+    {
+        if (!Arr::has($params, 'registrant.register')) {
+            throw $this->errorResult('Registrant contact data is required!');
+        }
+
+        if (!Arr::has($params, 'tech.register')) {
+            throw $this->errorResult('Tech contact data is required!');
+        }
+
+        if (!Arr::has($params, 'admin.register')) {
+            throw $this->errorResult('Admin contact data is required!');
+        }
+
+        if (!Arr::has($params, 'billing.register')) {
+            throw $this->errorResult('Billing contact data is required!');
+        }
+    }
+
+
+    /**
+     * @param ArrayAccess $params
+     * @param string $prefix
+     * @return string
+     */
+    private function prepareNameservers(ArrayAccess $params, string $prefix): string
+    {
+        $nameServers = "";
+
+        $custom = 0;
+        $default = 0;
+
+        for ($i = 1; $i <= self::MAX_CUSTOM_NAMESERVERS; $i++) {
+            if (Arr::has($params, $prefix.$i)) {
+                $host = Arr::get($params, $prefix.$i)->host;
+                if (!in_array($host, self::DEFAULT_NAMESERVERS)) {
+                    $nameServers .= $host.",";
+                    $custom++;
+                } else {
+                    $default++;
+                }
+            }
+        }
+
+        if ($custom != 0 && $default != 0) {
+            throw $this->errorResult(
+                "It's not possible to mix Namecheap's default nameservers with other ones",
```

Comment:
Please don't surface the Namecheap brand name in error messages- just refer to them as the "provider", since domain registrar services are typically re-sold white-label by web hosts.